### DISCUSSION TITLE
Backward chaining

### DISCUSSION THREAD

**Discussion by RekikMohamed** [NRules/NRules#365] [General]
Hello, 

We are trying to implement rule engine in our  application. We decided on nrules.
However, after research, we found out that nrules does not support backward chaining. Is there a way to support/simulate backward chaining in nrules? 
Are there any plans to add backward chaining to the nrules engine

**Comment by snikolayev** [comment]
Hello,

Indeed, there is no backward chaining capability in NRules. Implementing it is a huge amount of work. So, even though theoretically it is on the radar and is something that would be a great addition to the engine, realistically, it's not going to happen any time soon. Therefore, you should plan your project assuming it's not there.

**Comment by RekikMohamed** [comment]
@snikolayev  Thank you for your reply. 
Per my research, I found that we can simulate backward chaining with a workaround.  This example demonstrates how to check for loan approval through a query. The query operates in backward chaining mode, determining the approval based on existing facts.

This my fact: 
`public class LoanApplication
{
    public int CreditScore { get; set; }
    public bool Debt { get; set; }
    public bool Approved { get; set; }
}`

And this is the query: 
`using NRules;
using NRules.Fluent.Dsl;

public class ApproveLoanRule : Rule
{
    public override string Name => "Approve Loan";

    public override void Define()
    {
        When()
            .Match<LoanApplication>("application", a => a.CreditScore > 700 && !a.Debt);
        Then()
            .Do(ctx => 
            {
                var application = ctx.Get<LoanApplication>("application");
                application.Approved = true; // Approve the loan
            });
    }
}

public class CheckLoanApprovalRule : Rule
{
    public override string Name => "Check Loan Approval";

    public override void Define()
    {
        When()
            .Match<LoanApplication>("application", a => a.Approved);
        Then()
            .Do(ctx =>
            {
                var application = ctx.Get<LoanApplication>("application");
                // Logic that simulates a query, e.g., logging the approval
                Console.WriteLine("Loan is approved for the application.");
            });
    }
}`

`var session = factory.CreateSession();

var application = new LoanApplication { CreditScore = 750, Debt = false };
session.Insert(application);
session.Fire(); // The loan will be approved

// Check loan approval through the simulated backward chaining rule
session.Fire();`

Do you think this is a good idea to do it with nrules?

**Reply by snikolayev** [reply]
Sorry, I don't understand the idea, can you please elaborate?
Also, have you tried compiling your example?

**Reply by RekikMohamed** [reply]
@snikolayev  I compile the example and it works.

This example demonstrates how to check for loan approval through a query. The query operates in backward chaining mode, determining the approval based on existing facts.

The example above demonstrates how I implmeneted using NRules simulating backward chaining.

**Reply by snikolayev** [reply]
@RekikMohamed in your example you have one rule that updates the loan as approved, and another rule matching an approved loan. This is not backward chaining, this is forward chaining, so I don’t understand in which way it simulates backward chaining. I don’t think it does. If you think I’m missing something, please explain.
I also don’t see how this code can possibly compile. Match method does not accept string as the first argument; IContext doesn’t have a Get method, just to name a few things that won’t compile. Maybe you created some custom extension methods that do these things. In this case, please post a link to a repo with a compilable example, so that i can check.

**Reply by RekikMohamed** [reply]
@snikolayev  sorry for the confusion. It was my mistake. The code was not compiled. As per my understanding, backward chaining is to find facts that verifies a goal. The goal here is isLoanApproved. We get the value of isLoanApproved and we check does the fact that lead to is exists or not.

**Reply by snikolayev** [reply]
The purpose of both forward chaining and backward chaining is the same - to infer consequences from facts. The difference is in how the engine does it. In forward chaining it goes from facts to goals and consequences, in backward chaining it does it the other way around. So, it's different how many facts it needs to evaluate. One works well for some cases, while the other one works better for others, the difference would be the efficiency.
Imagine two hypothetical rules similar to what you wrote:
1. If LoanApplication has CreditScore > 700 and no debt => approve the loan
2. If LoanApplication is approved and applicant name is "RekikMohamed" => print the loan

Imagine also that the system has thousands of loan applications. A backward chaining engine would determine that it only needs to find an approved loan for one particular applicant, then it would only approve that one loan and print it. A forward chaining engine will approve all thousands of loans that satisfy the conditions, while still only printing one.
The outcome is the same, but in this case backward chaining is orders of magnitude more efficient. To make forward chaining as efficient you either need to only insert that one loan in the system (i.e. artificially limit what facts are available to the engine), or update all rules to be aware of which applicant you are looking for (explicitly embed intermediate goals to upstream rules).
NRules is a forward chaining engine, so it will behave the way I described. You cannot "simulate" backward chaining w/o implementing it in the engine, because it's about internal efficiency and order of evaluation, not about functional behavior.
So, when you are thinking about migrating your system to NRules, think about your rules and use cases, and whether your use cases lend themselves better to one or the other approach. I suggest you do a POC implementation and check the performance.
I hope this answers your original question.

On a separate note, moving forward, please only submit code examples that you verified compile and work, to be respectful of other people's time.

**Reply by RekikMohamed** [reply]
@snikolayev Thank you very much for your inputs and help. I will check and close the discussion for now. If there are any more issues I will re-open the discussion or create another one.
Thank you again.
